### ISSUE TITLE
'pack' Padding  does not run with LineBox

### ISSUE THREAD

**Issue by mdtrooper** [urwid/urwid#346]
##### Description:
LineBox fails into a Padding

##### Affected versions (if applicable)
```
$ dpkg -l | grep urwid
ii  python3-urwid                                               2.0.1-2+b1                                               amd64        curses-based UI/widget library for Python 3
```

##### Steps to reproduce (if applicable)
```
import urwid

palette = [
    ('a',         'black',      'dark cyan', 'standout'),
    ('b',       'white',      'dark red',   'standout')]

a = urwid.Text("AAA\nCCC\nDDD")
size = a.pack()
a = urwid.AttrWrap(a, 'a')
aa =urwid.LineBox(a)
aa = urwid.AttrWrap(aa, 'b')

# Bug
# ~ aaa = urwid.Filler(urwid.Padding(aa, 'center', 'pack'))

aaa = urwid.Filler(urwid.Padding(aa, 'center', size[0] + 2))

# ~ aaa = urwid.Filler(urwid.Padding(a, 'center', 'pack'))


urwid.MainLoop(aaa, palette).run()
```

##### Expected/actual outcome
[...]

**Comment by mdtrooper** [comment]
I want to help it...but I don't know where start to read the code and trying to fix.

I have start to decoration.py:419 class Padding(WidgetDecoration) .

**Comment by tonycpsu** [comment]
@mdtrooper , could you describe exactly what the bug is?

**Comment by mdtrooper** [comment]
Hi @tonycpsu , I found a bug with the third parameter of padding, the width. Because the documentation says:
http://urwid.org/reference/widget.html?highlight=padding#padding
> width – one of:
> given width
>     integer number of columns for self.original_widget
> 'pack'
>     try to pack self.original_widget to its ideal size
> ('relative', percentage of total width)
>     make width depend on the container’s width
> 'clip'
>     to enable clipping mode for a fixed widget "

And I wrote a code (with the min lines) that it shows the bug, you can uncomment the line 14 and you can see something to this:

![urwid padding_bug_pack](https://user-images.githubusercontent.com/1285451/57837227-ab21f180-77c2-11e9-9f16-3214572e4e3f.png)

And I have a workaround (I get the size manually of widget and set this size) and the padding works fine (as must working with 'pack'), with the 14 line commented and 16 line uncommented:

![urwid padding_workaround_bug_pack](https://user-images.githubusercontent.com/1285451/57837367-f6d49b00-77c2-11e9-9a85-b09f7eefc967.png)

Basically, the Padding (..., width='pack') does not run.

Regards.

**Comment by mdtrooper** [comment]
Sorry the image of bug has a green border and this green border is made with gimp (I clicked by error some button in the palette).

The correct image is:

![urwid padding_bug_pack](https://user-images.githubusercontent.com/1285451/57837763-b6c1e800-77c3-11e9-85c6-e2e9f747f4ce.png)

**Comment by ulidtko** [comment]
@tonycpsu just run the repro script. The only way it renders "compactly" is with manual sizing, `Padding(..., size[0] + 2)`, which is also the expected result for `Padding(... width='pack')`.

`width='pack'` however, renders the linebox child at full-width. Which is incorrect.

@mdtrooper please, clean up the extra screenshots you uploaded by mistake, by editing your comments, thanks. All that noise makes the issue harder to understand. Regarding where to start reading code: you're right, `decoration.py` is the relevant place, continue your analysis there.

**Comment by mdtrooper** [comment]
Well I fixed some months ago. Could check it?
